### ISSUE TITLE
Parsing utf-8 strings followed by garbage

### ISSUE THREAD

**Issue by ckrooss** [construct/construct#307]
Hi,  
I'm trying to parse a unicode string that is followed by random memory (remote system I don't control).
I can't seem to find a way to parse everything up to the \x00 and stop trying to parse after that. The width of the string field is supposed to be fixed at 8, \x00 indicates the end of the string.

Example:
```python
from construct import String, Struct

myStruct = Struct("name" / String(9, encoding="utf-8"))

raw_data = b"ABC\x00\x00\xde\xad\xc0\xde"
print(myStruct.parse(raw_data))
```
raises `UnicodeDecodeError: 'utf-8' codec can't decode byte 0xc0 in position 7: invalid start byte`.

`b"ABC\x00\x00\x00\x00\x00\x00"` works of course, because the \x00s are correctly trimmed.

Is there any way to parse utf-8 strings until \x00 occurs? I have worked around this by using Array(9, Int8) and running

```python
name = []
for c in parsed_data.name:
    if c == 0:
        break
    name.append(c)

parsed_data.name = bytes(name).decode("utf-8")
```
But this seems rather inelegant compared to using construct directly.
So TL;DR: is there currently a way to do this and if not how/where would I start adding it to construct?

**Comment by arekbulski** [comment]
I would suggest moving that small helper code into an Adapter. As for the bug, still trying to understand what goes wrong there.

**Comment by arekbulski** [comment]
Can you add a unit test with any string literal that fails this way? Would be huge helper. Please submit as a PR.

**Comment by ckrooss** [comment]
As far as I understand the construct code, it expects a string to be completely valid utf-8 when parsing. So if you have a String(10), the data must contain 10 bytes that are valid utf-8. My use-case is a String(10) that is **at most** 10 bytes long, but possibly shorter with trailing garbage bytes.  
What I would expect:
```python
String(3, encoding="utf8").parse(b"a\x00\x00") == u"a"
String(3, encoding="utf8").parse(b"a\x00\xFF") == u"a"
```
Please check the referenced PR if this is what you meant.

**Comment by arekbulski** [comment]
This is a very proper PR, thank you. It seems to me like what you are requesting is changing a proper code to work just so for this particular case. Yeah, the code assumes the bytes to be a proper UTF8 byte representation. 

I would suggest this: we add an adapter to the framework that finds the location of first zero byte, truncates the string and uses a subcon to parse what is left. Would you submit such a piece as PR?

**Comment by ckrooss** [comment]
Sure, that sounds like a perfect solution for me.
What would a idiomatic adapter look like? This seems to work for me, not sure if it fits the libraries style:

```python
def TrimUnicodeString(encoding, subcon):
    r"""
    Read bytes from Int8 array until \x00 is encountered
    Ignore everything after the first \x00
    Parse the bytes before \x00 and return decoded string

    Usage: TrimCString(encoding, bytes)
    Example::

        >>> TrimCString("utf-8", Array(10, Int8ul)).parse(b"ABC\x00\x00\xde\xad\xc0\xde\x00")
        u"ABC"

    """

    def decode(obj, context):
        output = []
        for c in obj:
            if c == 0:
                break
            output.append(c)
        return bytes(output).decode(encoding)

    def encode(obj, context):
        r"""
            Note: encode uses the original subcon length
            instead of the stripped output length
        """
        return String(length=subcon.sizeof(), encoding=encoding).build(obj)

    return ExprAdapter(subcon, encoder=encode, decoder=decode)
```

In this adapter,
decode returns a `<class 'str'>` instead of `String(...).parse(output)`, is that ok?

Also encode(decode()) does not return the original object because the trailing bytes are removed. Is that generally acceptable?

**Comment by arekbulski** [comment]
Rename to TrimString, as this would work on ASCII and bytes just as well. Drop macro in favor of just a class deriving from an Adapter. Make decode operate on bytes only, just find first index of zero and slice the bytes, then pass to subcon for parsing. If no zero then pass entire. Drop encode method.

**Comment by ckrooss** [comment]
I can't seem to get the correct place to do the slicing.
When I do 
```python
myStr = TrimString(String(length=10, encoding="utf-8"))
TrimString.parse(raw_data)
``` 
TrimString._parse calls subcon._parse which calls subcon._decode which is StringPaddedTrimmed._decode. At this point, it runs `obj.encode(encoding)` and raises UnicodeDecodeError.
I'm afraid I'll have to live with the function above until I can spend some more time with this.

**Comment by arekbulski** [comment]
Add your current code to the PR, just for the record.

**Comment by arekbulski** [comment]
Why not just create a new (fifth?) kind of string? TrimString(10,encoding).

**Comment by ckrooss** [comment]
I'm currently living with a filter that does the job but is probably not merge-quality. Would you suggest I add a TrimString in core.py similar to CString and PascalString?

**Comment by arekbulski** [comment]
I guess then just PR what you have. Even if I reject it, it will still remain in the repository for reference.
